runtime/src/main/java/org/corfudb/util/concurrent/SingletonResource.java

```diff
@@ -66,13 +75,10 @@ public T get() {
      * @param cleaner A {@link Consumer} which is provided the resource to perform cleanup
      *                actions.
      */
-    public void cleanup(@Nonnull Consumer<T> cleaner) {
-        resource.updateAndGet(t -> {
-            if (t != null) {
-                cleaner.accept(t);
-            }
-            return null;
-        });
+    public synchronized void cleanup(@Nonnull Consumer<T> cleaner) {
+        if (resource != null) {
+            cleaner.accept(resource);
```

Comment:
Set it back to null if we want the object to be reusable.

Reply:
see @StrikingThirteen 's comment.   I think its best we remove this method, it doesn't make a lot of sense.

Reply:
Rather than removing, we can change the name to destroy. We do need to clean up the resource - in this case, shut down the runtime.
Once destroyed the resource cannot be recreated/re-instantiated.

Reply:
> We do need to clean up the resource - in this case

I understand, the last paragraph of my comment says:

> The resource should be cleaned up externally. In other words, acquire the resource and then call clean up on it.

The Singleton itself doesn't need to have a destroy, shutdown should be called on the object after it is acquired. Trying to enforce a single call to  "close/shutdown" through a singleton just doesn't seem right.

Reply:
Ok, that sounds good.

Reply:
If we do null out the resource the race problem that I mentioned in the other comment will be present. If we don't, probably there is no point of keeping the cleanup around.